### PR TITLE
dbt-materialize: Catalog Fetch

### PR DESCRIPTION
**Pull request by dehume**
### Motivation

As part of the dbt v1.7.0 upgrade. Decomposing the `materialize__get_catalog` macro into multiple macros to allow for more performant querying.

### Tips for reviewer
This will fix the existing tests that are introduced with the upgrade to v1.7.0.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by morsapaes** [comment]
This PR covers a couple of different changes shipping with dbt v.1.7.0. Listing them out for clarity and shared context:

#### Testing relation types generated in dbt docs
The new `relation_types` test checks that the correct object type for each model is displayed in the dbt docs. This was motivated by a bunch of adapters not properly reporting the `table_type` of materialized views (and dynamic tables) in the `catalog.sql` macro (including [`dbt-postgres`](https://github.com/dbt-labs/dbt-core/pull/8945/files)).

The test has been exposed to the adapter framework in [dbt Core #8952](https://github.com/dbt-labs/dbt-core/pull/8952), so we'll be able to just override it in the next dbt release.

🎏 **TODO**
- [ ] Extend `relation_types` to test sources, subsources, and sinks.
- [ ] (Nice-to-have) Make `table_type` style consistent with other adapters in `catalog.sql` (_e.g._ [`materializedview`](https://github.com/MaterializeInc/materialize/blob/fe2eba966b967b543a6160e8740046c5e3cce1f3/misc/dbt-materialize/dbt/include/materialize/macros/catalog.sql#L27) -> [`MATERIALIZED VIEW`](https://github.com/dbt-labs/dbt-core/blob/e24f9b3da722328e24ee10a06fa2e8d2ca73b3e6/plugins/postgres/dbt/include/postgres/macros/catalog.sql#L18)). It'd be nice to at least adjust `materializedview`->`materialized_view`.

#### Slimming down catalog metadata
Prior to v1.7.0, dbt would fetch metadata for all objects in all schemas touched by a dbt project, even if these weren't used in the scope of the project. This release breaks down `get_catalog()` into multiple macros that reduce the amount of metadata retrieved, which helps speed up docs generation for smaller ([> 100 models](https://github.com/dbt-labs/dbt-core/blob/e24f9b3da722328e24ee10a06fa2e8d2ca73b3e6/core/dbt/adapters/base/impl.py#L1168)) projects.

Is there a particular reason you followed the [`dbt-snowflake` implementation](https://github.com/dbt-labs/dbt-snowflake/pull/758/files) instead of the [`dbt-postgres` one](https://github.com/dbt-labs/dbt-core/pull/8648/files#diff-094b5b5d15bb7e12be90a891c4e49184984d427d7b4657980068a7cad46b6512) here, @dehume? Can't put my finger on the difference, from a quick skim.

🎏 **TODO**

- [ ] Add [`TestDocsGenerateOverride`](https://github.com/dbt-labs/dbt-core/blob/main/tests/functional/artifacts/test_override.py) test.

#### Configuring feature support per-adapter
Ah! This could give us a nice little break from polluting the codebase with macro overrides to just declare a feature as unsupported or unimplemented. `dbt-postgres` already implements this in [`impl.py`](https://github.com/dbt-labs/dbt-core/blob/e24f9b3da722328e24ee10a06fa2e8d2ca73b3e6/plugins/postgres/dbt/adapters/postgres/impl.py#L79), though, so there is no need to change `impl.py` on our end.

**Comment by morsapaes** [comment]
Reading [this comment](https://github.com/dbt-labs/dbt-core/discussions/8307#discussioncomment-7325742) makes me uneasy about shipping the catalog metadata change now, since there is no real adapter-specific test coverage. We don't yet have users with an absurd amount of models, so we could kick this to a future release.

**Comment by dehume** [comment]
> Is there a particular reason you followed the [`dbt-snowflake` implementation](https://github.com/dbt-labs/dbt-snowflake/pull/758/files) instead of the [`dbt-postgres` one](https://github.com/dbt-labs/dbt-core/pull/8648/files#diff-094b5b5d15bb7e12be90a891c4e49184984d427d7b4657980068a7cad46b6512) here, @dehume? Can't put my finger on the difference, from a quick skim.

The original issue linked to the Snowflake implementation so based it on that. It is more verbose compared to the postgres but behaves the same.

File: misc/dbt-materialize/tests/adapter/test_relation_types.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        "node_name,relation_type",
+        [
+            ("seed.test.test_seed", "table"),
+            # NOTE(dehume): Tables are materialized as materialized views
+            # https://github.com/MaterializeInc/materialize/issues/5266
+            ("model.test.test_table", "materialized-view"),
+            ("model.test.test_view", "view"),
+            ("model.test.test_materialized_view", "materialized-view"),
+            ("model.test.test_source", "kafka"),
+            # ("model.test.test_subsource", "subsource"),
```
Comment (dehume) [inline misc/dbt-materialize/tests/adapter/test_relation_types.py:68]: While the dbt run is successful, it is not picking up the `subsource` or `sink` models. Running the catalog query on an environment with subsources and sinks does return those objects. Not sure if there is something with the test implementation that is missing.

Comment (morsapaes) [reply misc/dbt-materialize/tests/adapter/test_relation_types.py]: Reverting the extension to see if the test passes for the basic types. I'll try extending it after that.

File: misc/dbt-materialize/dbt/include/materialize/macros/catalog.sql [hunk 2]
```diff
@@ -49,6 +50,6 @@
                 '{{ schema }}' {%- if not loop.last %}, {% endif -%}
             {%- endfor -%}
         )
-  {%- endcall -%}
-  {{ return(load_result('catalog').table) }}
+    {%- endcall -%}
+    {{ return(load_result('catalog').table) }}
```
Comment (morsapaes) [inline misc/dbt-materialize/dbt/include/materialize/macros/catalog.sql]:
```suggestion
  {%- endcall -%}
  {{ return(load_result('catalog').table) }}
```

**Review by morsapaes — APPROVED** [review]

**Pull request merged**